runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt

```diff
@@ -92,10 +92,26 @@ public class FlexibleChecksumsRequestInterceptor<I>(
         } else if (req.headers[headerName] == null) {
             logger.debug { "Calculating checksum" }
 
-            val bodyBytes = req.body.readAll()!!
-            req.body = bodyBytes.toHttpBody() // replace the consumed body
+            val checksum: String = when {
+                req.body.contentLength == null && !req.body.isOneShot -> {
+                    // calculate the checksum using a rolling hash
+                    val buffer = SdkBuffer()
+                    val bufferSize: Long = 8192
+
+                    val channel = req.body.toSdkByteReadChannel()!!
+                    while (!channel.isClosedForRead) {
+                        channel.read(buffer, bufferSize)
+                        checksumAlgorithm.update(buffer.readToByteArray())
+                    }
+                    checksumAlgorithm.digest().encodeBase64String()
```

Comment:
**Style**: I suggest moving the checksum calculation to its own function (do we already have an extension function somewhere for this?) and only encoding the digest to Base64 in this method.